crypto/types/compact_bit_array.go

```diff
@@ -94,14 +95,13 @@ func (bA *CompactBitArray) NumTrueBitsBefore(index int) int {
 	var i = 0
 	for elem := 0; ; elem++ {
 		for b := 7; b >= 0; b-- {
-			if i >= index {
-				return numTrueValues
+			if i == index {
+				onesCount += bits.OnesCount8(bA.Elems[elem] >> (b + 1))
+				return onesCount
 			}
 			i++
-			if (bA.Elems[elem]>>b)&1 == 1 {
-				numTrueValues++
-			}
 		}
```

Comment:
After the update, the inner loop is not needed. Instead you can do:
```
for elem := 0; ; elem++ {
  if elem*8 + 7 >= index {   
    onesCount += bits.OnesCount8(bA.Elems[elem] >> (b + 1))
    return onesCount
  }
  onesCount += bits.OnesCount8(bA.Elems[elem])
}

Reply:
Thanks, the performance gain is double comparing when keeping the inner loop. I think we should refactor all the `elem*8+7` and `7 - (index&8)` to make the code clearer. If you think it's ok, I will do it in other PR.

Reply:
This loop is updated in this PR, so let's clean that loop in this PR. It's only 3 lines of code to change - it's weird living a useless inner loop here. You can refactor other index arithmetic in other PR.

Reply:
@robert-zaremba Yes, I did. To clarify, what I mean is that removing the inner loop gain double the performance comparing with keeping it. PTAL!